src/DotNetWorker.Grpc/FunctionMetadata/FunctionMetadataJsonReader.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading.Tasks;
+using System.Text.Json;
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+
+namespace Microsoft.Azure.Functions.Worker
+{ 
+    internal class FunctionMetadataJsonReader
+    {
+        private readonly string _directory;
+        private const string FileName = "functions.metadata";
+
+        public FunctionMetadataJsonReader(string directory)
+        {
+            _directory = directory ?? throw new ArgumentNullException(nameof(directory));
+        }
+
+        public virtual async Task<ImmutableArray<FunctionLoadRequest>> ReadMetadataAsync()
+        {
+            string metadataFile = Path.Combine(_directory, FileName);
+
+            if (File.Exists(metadataFile))
+            {
+                using (var fs = File.OpenRead(metadataFile))
+                {
+                    var jsonMetadataList = await JsonSerializer.DeserializeAsync<JsonElement>(fs);
+
+                    var functionRequests = new List<FunctionLoadRequest>(jsonMetadataList.GetArrayLength());
+
+                    foreach (var jsonMetadata in jsonMetadataList.EnumerateArray())
+                    {
+                        var options = new JsonSerializerOptions();
+                        options.PropertyNameCaseInsensitive = true;
+
+                        var functionMetadata = JsonSerializer.Deserialize<RpcFunctionMetadata>(jsonMetadata.GetRawText(), options);
```

Comment:
There is an [async version of this method](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserializeasync?view=net-5.0#System_Text_Json_JsonSerializer_DeserializeAsync__1_System_IO_Stream_System_Text_Json_JsonSerializerOptions_System_Threading_CancellationToken_) (`DeserializeAsync`) method you could consider switching to.

Reply:
Also, I wonder whether we really need to convert the `JsonElement` to `string`, then deserialize it again. Could we not read the prop values from the `JsonElement` instance and set the properties of `RpcFunctionMetadata` instance manually?

Reply:
I tried it with GetProperty() and TryGetProperty() first, but that leads to null-checking every property when we grab them and call ToString().

Didn't use the async version b/c it throws an error: `'RpcFunctionMetadata' does not contain a definition for 'GetAwaiter' ... `

Reply:
Were you awaiting it? Line 32 uses that overload. So should work here as well.

Reply:
Shyju and I just discussed this in a call, but since we have already read from the file stream to get the raw json (and preserve the raw bindings) we would need to create a new stream for the jsonMetadata. This class and its methods are used only once at initialization to provide function metadata to the host, so there are not many advantages to switch to an async version at the moment.